utxo/src/utxo_impl/mod.rs

```diff
@@ -0,0 +1,537 @@
+                }
+            }
+        };
+
+        // create a new entry
+        let new_entry = UtxoEntry::new(utxo, is_fresh, true);
+
+        // TODO: update the memory usage
+        // self.memory_usage should be added based on this new entry.
+
+        self.utxos.insert(key.clone(), new_entry);
+
+        Ok(())
+    }
+
+    /// Flags the utxo as "spent", given an outpoint.
+    /// Returns the Utxo if an update was performed.
+    pub fn spend_utxo(&mut self, outpoint: &OutPoint) -> Result<Utxo, Error> {
+        match self.get_utxo_entry(outpoint) {
+            None => Err(Error::NoUtxoFound),
+            Some(entry) => {
+                let key = outpoint;
+
+                // TODO: update the memory usage
+                // self.memory_usage must be deducted from this entry's size
+
+                // check whether this entry is fresh
+                if entry.is_fresh {
+                    // This is only available in this view. Remove immediately.
+                    self.utxos.remove(key);
+                } else {
+                    // mark this as 'spent'
+                    let new_entry = UtxoEntry {
+                        status: UtxoStatus::Spent,
+                        is_dirty: true,
+                        is_fresh: false,
+                    };
+                    self.utxos.insert(key.clone(), new_entry);
+                }
+
+                entry.take_utxo().ok_or(Error::UtxoAlreadySpent)
+            }
+        }
+    }
+
+    /// Checks whether utxo exists in the cache
+    pub fn has_utxo_in_cache(&self, outpoint: &OutPoint) -> bool {
+        let key = outpoint;
+        self.utxos.contains_key(key)
+    }
+
+    /// Returns a mutable reference of the utxo, given the outpoint.
+    pub fn get_mut_utxo(&mut self, outpoint: &OutPoint) -> Option<&mut Utxo> {
+        self.get_utxo_entry(outpoint).and_then(|status| {
+            match status.status {
+                UtxoStatus::Spent => None,
+                UtxoStatus::Entry(utxo) => {
+                    let key = outpoint;
+                    self.utxos.insert(
+                        key.clone(),
+                        UtxoEntry::new(utxo, status.is_fresh, status.is_dirty),
+                    );
+                    //TODO: update the memory storage here
+                    self.utxos.get_mut(key).and_then(|entry| entry.utxo_mut())
+                }
+            }
+        })
+    }
+
+    /// removes the utxo in the cache with the outpoint
+    pub(crate) fn uncache(&mut self, outpoint: &OutPoint) -> Option<UtxoEntry> {
+        let key = outpoint;
+        if let Some(entry) = self.utxos.get(key) {
+            // see bitcoin's Uncache.
+            if !entry.is_fresh && !entry.is_dirty {
+                //todo: decrement the memory usage
+                return self.utxos.remove(key);
+            }
+        }
+        None
+    }
+
+    fn consume(self) -> Result<ConsumedUtxoCache, Error> {
```

Comment:
This is a thinker. I'm really not sure about this one. While I don't really care whether it's finalize or consume, I feel like FinalUtxoCache doesn't really convey what's going on. Unlike crypto stream hasher, where a finalize is understandable to get the result of hashing. I'm happy to hear more opinions about this. I use the same terminology in the Consensus crate, btw for the reason with spent inputs cache.

Reply:
Ok I get it and don't have a strong opinion. It just seems to me that `consume` conveys more about implementation details rather than the essence of the operation.